discovery/discovery.go

```diff
@@ -0,0 +1,302 @@
+		k, err := kubernetes.New(log.Base(), c)
+		if err != nil {
+			log.Errorf("Cannot create Kubernetes discovery: %s", err)
+			continue
+		}
+		app("kubernetes", i, k)
+	}
+	for i, c := range cfg.ServersetSDConfigs {
+		app("serverset", i, zookeeper.NewServersetDiscovery(c))
+	}
+	for i, c := range cfg.NerveSDConfigs {
+		app("nerve", i, zookeeper.NewNerveDiscovery(c))
+	}
+	for i, c := range cfg.EC2SDConfigs {
+		app("ec2", i, ec2.NewDiscovery(c))
+	}
+	for i, c := range cfg.GCESDConfigs {
+		gced, err := gce.NewDiscovery(c)
+		if err != nil {
+			log.Errorf("Cannot initialize GCE discovery: %s", err)
+			continue
+		}
+		app("gce", i, gced)
+	}
+	for i, c := range cfg.AzureSDConfigs {
+		app("azure", i, azure.NewDiscovery(c))
+	}
+	if len(cfg.StaticConfigs) > 0 {
+		app("static", 0, NewStaticProvider(cfg.StaticConfigs))
+	}
+
+	return providers
+}
+
+// StaticProvider holds a list of target groups that never change.
+type StaticProvider struct {
+	TargetGroups []*config.TargetGroup
+}
+
+// NewStaticProvider returns a StaticProvider configured with the given
+// target groups.
+func NewStaticProvider(groups []*config.TargetGroup) *StaticProvider {
+	for i, tg := range groups {
+		tg.Source = fmt.Sprintf("%d", i)
+	}
+	return &StaticProvider{groups}
+}
+
+// Run implements the TargetProvider interface.
+func (sd *StaticProvider) Run(ctx context.Context, ch chan<- []*config.TargetGroup) {
+	// We still have to consider that the consumer exits right away in which case
+	// the context will be canceled.
+	select {
+	case ch <- sd.TargetGroups:
+	case <-ctx.Done():
+	}
+	close(ch)
+}
+
+// TargetSet handles multiple TargetProviders and sends a full overview of their
+// discovered TargetGroups to a Syncer.
+type TargetSet struct {
+	mtx sync.RWMutex
+	// Sets of targets by a source string that is unique across target providers.
+	tgroups map[string]*config.TargetGroup
+
+	syncer Syncer
+
+	syncCh          chan struct{}
+	providerCh      chan map[string]TargetProvider
+	cancelProviders func()
+}
+
+// Syncer receives updates complete sets of TargetGroups.
+type Syncer interface {
+	Sync([]*config.TargetGroup)
+}
+
+// NewTargetSet reutrns a new target sending TargetGroups to the syncer.
+func NewTargetSet(s Syncer) *TargetSet {
+	return &TargetSet{
+		syncCh:     make(chan struct{}, 1),
+		providerCh: make(chan map[string]TargetProvider),
+		syncer:     s,
+	}
+}
+
+// Run starts the processing of target providers and their updates.
+// It blocks until the context gets canceled.
+func (ts *TargetSet) Run(ctx context.Context) {
+Loop:
+	for {
+		// Throttle syncing to once per five seconds.
+		select {
+		case <-ctx.Done():
+			break Loop
+		case p := <-ts.providerCh:
+			ts.updateProviders(ctx, p)
+		case <-time.After(5 * time.Second):
+		}
+
+		select {
+		case <-ctx.Done():
+			break Loop
+		case <-ts.syncCh:
+			ts.sync()
+		case p := <-ts.providerCh:
+			ts.updateProviders(ctx, p)
+		}
+	}
+}
+
+func (ts *TargetSet) sync() {
+	ts.mtx.RLock()
```

Comment:
Should we drain ts.syncCh here in case several have backed up? Having to wait 5s for every pending update would be ungood.